File: tools/replay_har_consulta_placa.py
#!/usr/bin/env python3
import argparse
import base64
import json
import os
import re
import sys
from typing import Any, Dict, Optional, Tuple

import requests


SENSITIVE_HEADERS = {"authorization", "cookie", "set-cookie", "proxy-authorization"}


def _redact_headers(headers: Dict[str, str], show_secrets: bool) -> Dict[str, str]:
    if show_secrets:
        return dict(headers)
    redacted: Dict[str, str] = {}
    for name, value in headers.items():
        if name.lower() in SENSITIVE_HEADERS and value:
            redacted[name] = "[REDACTED]"
        else:
            redacted[name] = value
    return redacted


def _har_headers_to_dict(har_headers: Any) -> Dict[str, str]:
    out: Dict[str, str] = {}
    for h in har_headers or []:
        name = h.get("name")
        value = h.get("value")
        if not name:
            continue
        if value is None:
            value = ""
        out[name] = value
    return out


def _strip_hop_by_hop(headers: Dict[str, str]) -> Dict[str, str]:
    hop_by_hop = {
        "connection",
        "host",
        "content-length",
        "accept-encoding",
        "proxy-connection",
        "transfer-encoding",
        "upgrade",
    }
    return {k: v for k, v in headers.items() if k.lower() not in hop_by_hop}


def _find_entry(
    har: Dict[str, Any],
    entry_index: Optional[int],
    url_regex: Optional[str],
) -> Tuple[int, Dict[str, Any]]:
    entries = har.get("log", {}).get("entries", []) or []
    if entry_index is not None:
        if entry_index < 0 or entry_index >= len(entries):
            raise SystemExit(f"--entry-index fora do range: 0..{len(entries) - 1}")
        return entry_index, entries[entry_index]

    pattern = re.compile(url_regex or r"ChkRetorno2HTML", re.IGNORECASE)
    for idx, entry in enumerate(entries):
        url = (entry.get("request", {}) or {}).get("url") or ""
        if pattern.search(url):
            return idx, entry

    raise SystemExit("Nenhuma entry encontrada (tente --url-regex ou --entry-index).")


def _decode_har_response_body(entry: Dict[str, Any]) -> Optional[str]:
    content = (entry.get("response", {}) or {}).get("content", {}) or {}
    text = content.get("text")
    if not text:
        return None

    encoding = content.get("encoding")
    if encoding == "base64":
        try:
            return base64.b64decode(text).decode("utf-8", errors="replace")
        except Exception:
            return base64.b64decode(text).decode("latin-1", errors="replace")
    return text


def _maybe_parse_json(text: str) -> Any:
    try:
        return json.loads(text)
    except Exception:
        return None


def main() -> int:
    ap = argparse.ArgumentParser(
        description=(
            "Analisa um HAR e (opcionalmente) faz replay da requisição de consulta "
            "(ex.: ChkRetorno2HTML), decodificando o retorno."
        )
    )
    ap.add_argument("--har", default="e-system/teste.har", help="Caminho do arquivo .har")
    ap.add_argument("--entry-index", type=int, default=None, help="Índice exato da entry no HAR")
    ap.add_argument("--url-regex", default=None, help="Regex para escolher a entry (default: ChkRetorno2HTML)")
    ap.add_argument(
        "--authorization",
        default=None,
        help=(
            "Header Authorization. Se omitido, usa $AUTHORIZATION. "
            "Evite depender do valor que estiver dentro do HAR."
        ),
    )
    ap.add_argument(
        "--override-param2",
        default=None,
        help="Substitui payload._parameters[2] (ex.: '|94DFAAP16TB015294').",
    )
    ap.add_argument("--show-secrets", action="store_true", help="Mostra headers sensíveis (cuidado).")
    ap.add_argument("--replay", action="store_true", help="Faz replay HTTP da entry selecionada.")
    ap.add_argument("--timeout", type=float, default=60.0, help="Timeout (segundos) para o replay.")
    args = ap.parse_args()

    with open(args.har, "rb") as f:
        har = json.load(f)

    idx, entry = _find_entry(har, args.entry_index, args.url_regex)
    req = entry.get("request", {}) or {}

    method = req.get("method") or "GET"
    url = req.get("url") or ""
    headers = _har_headers_to_dict(req.get("headers"))
    headers = _strip_hop_by_hop(headers)

    if args.authorization is None:
        args.authorization = os.environ.get("AUTHORIZATION")
    if args.authorization:
        headers["Authorization"] = args.authorization

    post_data = (req.get("postData", {}) or {}).get("text") or ""
    if args.override_param2 and post_data:
        try:
            payload = json.loads(post_data)
            params = payload.get("_parameters")
            if isinstance(params, list) and len(params) >= 3:
                params[2] = args.override_param2
                post_data = json.dumps(payload, ensure_ascii=False, separators=(",", ":"))
        except Exception:
            pass

    print(f"Entry: {idx}")
    print(f"Request: {method} {url}")
    print("Headers:", json.dumps(_redact_headers(headers, args.show_secrets), ensure_ascii=False, indent=2))

    decoded_har_body = _decode_har_response_body(entry)
    if decoded_har_body is not None:
        parsed = _maybe_parse_json(decoded_har_body)
        if parsed is not None:
            print("HAR response (decoded JSON):", json.dumps(parsed, ensure_ascii=False))
        else:
            print("HAR response (decoded):", decoded_har_body[:5000])

    if not args.replay:
        return 0

    if method.upper() != "POST":
        print("Replay só implementado para POST neste script. Use --no-replay ou escolha outra entry.")
        return 2

    mime = (req.get("postData", {}) or {}).get("mimeType")
    if mime and "content-type" not in {k.lower() for k in headers.keys()}:
        headers["Content-Type"] = mime

    resp = requests.request(
        method=method.upper(),
        url=url,
        headers=headers,
        data=post_data.encode("utf-8") if isinstance(post_data, str) else post_data,
        timeout=args.timeout,
    )

    print(f"HTTP status: {resp.status_code}")
    content_type = resp.headers.get("content-type", "")
    print(f"Content-Type: {content_type}")

    try:
        print("Replay response JSON:", json.dumps(resp.json(), ensure_ascii=False))
        return 0
    except Exception:
        body = resp.text
        # fallback: algumas APIs respondem base64; tenta decodificar
        try:
            decoded = base64.b64decode(body).decode("utf-8", errors="replace")
            parsed = _maybe_parse_json(decoded)
            if parsed is not None:
                print("Replay response (base64->JSON):", json.dumps(parsed, ensure_ascii=False))
                return 0
            print("Replay response (base64 decoded):", decoded[:5000])
            return 0
        except Exception:
            print("Replay response (text):", body[:5000])
            return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/replay_har_consulta_placa.mjs
#!/usr/bin/env node
import fs from "node:fs/promises";

const SENSITIVE_HEADERS = new Set([
  "authorization",
  "cookie",
  "set-cookie",
  "proxy-authorization",
]);

function parseArgs(argv) {
  const args = {};
  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i];
    if (!token.startsWith("--")) continue;
    const key = token.slice(2);
    const next = argv[i + 1];
    if (!next || next.startsWith("--")) {
      args[key] = true;
    } else {
      args[key] = next;
      i += 1;
    }
  }
  return args;
}

function harHeadersToObject(harHeaders) {
  const out = {};
  for (const h of harHeaders ?? []) {
    if (!h?.name) continue;
    out[h.name] = h.value ?? "";
  }
  return out;
}

function stripHopByHop(headers) {
  const hopByHop = new Set([
    "connection",
    "host",
    "content-length",
    "accept-encoding",
    "proxy-connection",
    "transfer-encoding",
    "upgrade",
  ]);

  const out = {};
  for (const [k, v] of Object.entries(headers)) {
    if (hopByHop.has(k.toLowerCase())) continue;
    out[k] = v;
  }
  return out;
}

function redactHeaders(headers, showSecrets) {
  if (showSecrets) return { ...headers };
  const out = {};
  for (const [k, v] of Object.entries(headers)) {
    if (SENSITIVE_HEADERS.has(k.toLowerCase()) && v) out[k] = "[REDACTED]";
    else out[k] = v;
  }
  return out;
}

function decodeHarResponseBody(entry) {
  const content = entry?.response?.content ?? {};
  const text = content.text;
  if (!text) return null;
  if (content.encoding === "base64") {
    return Buffer.from(text, "base64").toString("utf8");
  }
  return text;
}

function findEntry(entries, entryIndex, urlRegex) {
  if (entryIndex !== undefined) {
    const idx = Number(entryIndex);
    if (!Number.isFinite(idx) || idx < 0 || idx >= entries.length) {
      throw new Error(`--entry-index fora do range: 0..${entries.length - 1}`);
    }
    return { idx, entry: entries[idx] };
  }

  const pattern = new RegExp(urlRegex ?? "ChkRetorno2HTML", "i");
  for (let idx = 0; idx < entries.length; idx += 1) {
    const url = entries[idx]?.request?.url ?? "";
    if (pattern.test(url)) return { idx, entry: entries[idx] };
  }

  throw new Error("Nenhuma entry encontrada (tente --url-regex ou --entry-index).");
}

function bytesToString(bytes) {
  return Buffer.from(bytes).toString("utf8");
}

function stringToBytesArray(str) {
  return Array.from(Buffer.from(str, "utf8"));
}

function replaceParamValue(source, key, value) {
  // param0 inclui uma URL com querystring, então fazemos um replace simples.
  // Ex.: chassi=XXXXX  / captchaResponse=YYYYY
  const re = new RegExp(`(${key}=)([^&|]*)`, "g");
  if (!re.test(source)) return source;
  return source.replace(re, `$1${value}`);
}

function tryJson(text) {
  try {
    return JSON.parse(text);
  } catch {
    return null;
  }
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const harPath = args.har ?? "e-system/consulta_placa.har";
  const showSecrets = Boolean(args["show-secrets"]);
  const replay = Boolean(args.replay);

  const raw = await fs.readFile(harPath, "utf8");
  const har = JSON.parse(raw);
  const entries = har?.log?.entries ?? [];
  if (!Array.isArray(entries) || entries.length === 0) {
    throw new Error("HAR sem entries.");
  }

  const { idx, entry } = findEntry(entries, args["entry-index"], args["url-regex"]);
  const req = entry.request ?? {};

  const method = (req.method ?? "GET").toUpperCase();
  const url = req.url ?? "";

  let headers = stripHopByHop(harHeadersToObject(req.headers));
  const authorization = args.authorization ?? process.env.AUTHORIZATION;
  if (authorization) headers.Authorization = authorization;

  const harRespBody = decodeHarResponseBody(entry);
  if (harRespBody != null) {
    const parsed = tryJson(harRespBody);
    if (parsed) {
      process.stdout.write(`HAR response (decoded JSON): ${JSON.stringify(parsed)}\n`);
    } else {
      process.stdout.write(`HAR response (decoded): ${harRespBody.slice(0, 5000)}\n`);
    }
  }

  process.stdout.write(`Entry: ${idx}\n`);
  process.stdout.write(`Request: ${method} ${url}\n`);
  process.stdout.write(
    `Headers: ${JSON.stringify(redactHeaders(headers, showSecrets), null, 2)}\n`,
  );

  if (!replay) return;

  if (method !== "POST") {
    throw new Error("Replay só implementado para POST neste script.");
  }

  let bodyText = req?.postData?.text ?? "";
  if (!bodyText) {
    throw new Error("Entry não tem postData.text.");
  }

  let payload = null;
  try {
    payload = JSON.parse(bodyText);
  } catch {
    payload = null;
  }

  if (payload && Array.isArray(payload._parameters)) {
    const params = payload._parameters;

    // Opcional: override do chassi e/ou captcha no param0 (string em bytes)
    if (args.chassi && Array.isArray(params[0])) {
      const p0 = bytesToString(params[0]);
      const updated = replaceParamValue(p0, "chassi", args.chassi);
      params[0] = stringToBytesArray(updated);
    }

    if (args.captcha && Array.isArray(params[0])) {
      const p0 = bytesToString(params[0]);
      const updated = replaceParamValue(p0, "captchaResponse", args.captcha);
      params[0] = stringToBytesArray(updated);
    }

    // param2 normalmente é "|<chassi/vin>"
    if (args.chassi && typeof params[2] === "string") {
      params[2] = `|${args.chassi}`;
    }

    bodyText = JSON.stringify(payload);
  }

  const contentType =
    headers["Content-Type"] ??
    headers["content-type"] ??
    req?.postData?.mimeType ??
    "text/plain;charset=UTF-8";

  headers["Content-Type"] = contentType;

  const resp = await fetch(url, {
    method,
    headers,
    body: bodyText,
  });

  const respText = await resp.text();
  process.stdout.write(`HTTP status: ${resp.status}\n`);
  process.stdout.write(`Content-Type: ${resp.headers.get("content-type") ?? ""}\n`);

  // Alguns endpoints retornam JSON base64 via "text" — tenta decodificar se for o caso
  const parsedDirect = tryJson(respText);
  if (parsedDirect) {
    process.stdout.write(`Replay response JSON: ${JSON.stringify(parsedDirect)}\n`);
    return;
  }

  // fallback: tenta base64 -> JSON
  try {
    const decoded = Buffer.from(respText, "base64").toString("utf8");
    const parsed = tryJson(decoded);
    if (parsed) {
      process.stdout.write(`Replay response (base64->JSON): ${JSON.stringify(parsed)}\n`);
      return;
    }
  } catch {
    // ignore
  }

  process.stdout.write(`Replay response (text): ${respText.slice(0, 5000)}\n`);
}

main().catch((err) => {
  console.error(String(err?.stack ?? err));
  process.exitCode = 1;
});

